### PR TITLE
Add JSDoc for createhash

### PR DESCRIPTION
**Pull request by mohsen1**
WIP

### CODE REVIEW

File: lib/util/createHash.js [hunk 1]
```diff
@@ -45,24 +62,41 @@ class BulkUpdateDecorator {
 }
 
 /* istanbul ignore next */
+/** @implements {crypto.Hash} */
 class DebugHash {
 	constructor() {
+		/** @type {string} */
```
Comment (ooflorent) [inline lib/util/createHash.js]: You could add `@private` here

**Comment by TheLarkInn** [comment]
Fixed merge conflicts :-)

Heads up this fixes any issues with `/** @typedef {import(SomeModule)} SomeModule */`

File: lib/HotModuleReplacementPlugin.js [hunk 2]
```diff
@@ -13,6 +13,8 @@ const ConstDependency = require("./dependencies/ConstDependency");
 const NullFactory = require("./NullFactory");
 const ParserHelpers = require("./ParserHelpers");
 
+/** @typedef {typeof import("crypto")} crypto */
```
Comment (sokra) [inline lib/HotModuleReplacementPlugin.js]: Not needed, remove it.

File: lib/util/SortableSet.js [hunk 3]
```diff
@@ -1,19 +1,32 @@
 "use strict";
-//TODO: Make this a generic type
-//https://github.com/Microsoft/TypeScript/issues/23385
-//https://github.com/Microsoft/TypeScript/issues/23384
+
+/**
+ * A subset of Set that offers sorting functionality
+ * @template T
+ *
+ * @typedef {(a: T, b: T) => number} SortFunction
```
Comment (sokra) [inline lib/util/SortableSet.js]: invalid jsdoc. Move this into a `@callback` section

Comment (sokra) [reply lib/util/SortableSet.js]: also return type should be `-1 | 0 | 1`

File: lib/util/SortableSet.js [hunk 4]
```diff
@@ -1,19 +1,32 @@
...
+ */
 class SortableSet extends Set {
+	/**
+	 * Create a new sortable set
+	 * @param {IterableIterator<T>=} initialIterable The initial iterable value
+	 * @param {SortFunction=} defaultSort Default sorting function
+	 */
 	constructor(initialIterable, defaultSort) {
 		super(initialIterable);
+		/** @private @type {(a: T, b: T) => number}} */
```
Comment (sokra) [inline lib/util/SortableSet.js]: `SortFunction`

File: lib/util/SortableSet.js [hunk 5]
```diff
@@ -1,19 +1,32 @@
...
 		this._sortFn = defaultSort;
+		/** @private @type {(a: T, b: T) => number} | null} */
```
Comment (sokra) [inline lib/util/SortableSet.js]: `SortFunction`

File: lib/util/SortableSet.js [hunk 6]
```diff
@@ -50,16 +75,13 @@ class SortableSet extends Set {
...
-	 * @returns {void}
-	 */
 	sort() {
 		this.sortWith(this._sortFn);
 	}
 
 	/**
-	 * @param {Function} fn - function to calculate value
-	 * @returns {TODO} - returns result of fn(this), cached until set changes
+	 * @param {(instance: this) => TODO} fn function to calculate value
```
Comment (sokra) [inline lib/util/SortableSet.js]: `@callback` section

File: lib/util/createHash.js [hunk 7]
```diff
@@ -6,12 +6,24 @@
...
+	 * @param {crypto.Hash} hash Hashing algorithm
+	 */
 	constructor(hash) {
 		this.hash = hash;
 		this.buffer = "";
 	}
 
+	/**
+	 * Update hash
+	 * @param {any} data Data to update
```
Comment (sokra) [inline lib/util/createHash.js]: `string | Buffer`

File: lib/util/createHash.js [hunk 8]
```diff
@@ -45,24 +62,46 @@ class BulkUpdateDecorator {
...
+/** @implements {crypto.Hash} */
 class DebugHash {
 	constructor() {
+		/** @type {string} */
 		this.string = "";
 	}
 
+	/**
+	 * Update hash
+	 * @param {any} data Data to update
```
Comment (sokra) [inline lib/util/createHash.js]: `string|Buffer`

File: lib/util/createHash.js [hunk 9]
```diff
@@ -45,24 +62,46 @@ class BulkUpdateDecorator {
...
 	digest(encoding) {
 		return this.string.replace(/[^a-z0-9]+/gi, m =>
 			Buffer.from(m).toString("hex")
 		);
 	}
 }
 
+/**
+ * TODO: This is a constructor function, type it as such
+ * @typedef {any} Algorithm
```
Comment (sokra) [inline lib/util/createHash.js]: TODO

**Comment by mohsen1** [comment]
Lets merge #7513 first. I'll make another PR

**Pull request closed without merging**